ambry-store/src/main/java/com.github.ambry.store/Log.java

```diff
@@ -13,161 +13,380 @@
+    Utils.preAllocateFileIfNeeded(segmentFile, size);
+    remainingUnallocatedSegments--;
+    return segmentFile;
+  }
+
+  /**
+   * Frees the given {@link LogSegment} and its backing segment file.
+   * @param logSegment the {@link LogSegment} instance whose backing file needs to be freed.
+   * @throws IOException if there is any I/O error freeing the log segment.
+   */
+  private void free(LogSegment logSegment) throws IOException {
+    // TODO (DiskManager changes): This will actually return the segment to the DiskManager pool.
+    File segmentFile = logSegment.getView().getFirst();
+    logSegment.close();
+    if (!segmentFile.delete()) {
+      throw new IllegalStateException("Could not delete segment file: " + segmentFile.getAbsolutePath());
+    }
+    remainingUnallocatedSegments++;
+  }
+
+  /**
+   * Rolls the active log segment over if required. If rollover is required, a new segment is allocated.
+   * @param writeSize the size of the incoming write.
+   * @throws IllegalArgumentException if the {@code writeSize} is greater than a single segment's size
+   * @throws IllegalStateException if there is no more capacity in the log.
+   * @throws IOException if any I/O error occurred as part of ensuring capacity.
+   *
+   */
+  private void rollOverIfRequired(long writeSize) throws IOException {
+    if (activeSegment.getCapacityInBytes() - activeSegment.getEndOffset() < writeSize) {
+      ensureCapacity(writeSize);
+      // this cannot be null since capacity has either been ensured or has thrown.
+      LogSegment nextActiveSegment = segmentsByName.higherEntry(activeSegment.getName()).getValue();
+      logger.info("Rolling over writes to {} from {} on write of data of size {}. End offset was {} and capacity is {}",
+          nextActiveSegment.getName(), activeSegment.getName(), writeSize, activeSegment.getEndOffset(),
+          activeSegment.getCapacityInBytes());
+      activeSegment = nextActiveSegment;
     }
-    fileChannel.read(buffer, position);
   }
-}
 
+  /**
+   * Ensures that there is enough capacity for a write of size {@code writeSize} in the log. As a part of ensuring
+   * capacity, this function will also allocate more segments if required.
+   * @param writeSize the size of a subsequent write on the active log segment.
+   * @throws IllegalArgumentException if the {@code writeSize} is greater than a single segment's size
+   * @throws IllegalStateException if there no more capacity in the log.
+   * @throws IOException if any I/O error occurred as a part of ensuring capacity.
+   */
+  private void ensureCapacity(long writeSize) throws IOException {
+    if (remainingUnallocatedSegments == 0) {
+      metrics.overflowWriteError.inc();
+      throw new IllegalStateException(
+          "There is no more capacity left in [" + dataDir + "]. Max capacity is [" + capacityInBytes + "]");
+    }
+    // all segments are (should be) the same size.
+    long segmentCapacity = activeSegment.getCapacityInBytes();
```

Comment:
the actual_capacity is cleaner I think. Doing a `- HEADER_SIZE` is a specialized use case that is used by components that are very aware of how the log segment works.